File: 02-topics-pubsub/src/talker_listener_cpp/src/talker.cpp
// 第 02 课：发布者（C++ 版）。
// 运行: ros2 run talker_listener_cpp talker

#include <chrono>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

using namespace std::chrono_literals;

class Talker : public rclcpp::Node
{
public:
  Talker()
  : Node("talker"), count_(0)
  {
    publisher_ = this->create_publisher<std_msgs::msg::String>("chatter", 10);
    timer_ = this->create_wall_timer(1s, std::bind(&Talker::on_timer, this));
    RCLCPP_INFO(this->get_logger(), "Talker 已上线，正在向 /chatter 发布消息……");
  }

private:
  void on_timer()
  {
    auto msg = std_msgs::msg::String();
    msg.data = "Hello from talker #" + std::to_string(count_);
    publisher_->publish(msg);
    RCLCPP_INFO(this->get_logger(), "发布: %s", msg.data.c_str());
    ++count_;
  }

  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
  rclcpp::TimerBase::SharedPtr timer_;
  int count_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<Talker>());
  rclcpp::shutdown();
  return 0;
}

File: 02-topics-pubsub/src/talker_listener/talker_listener/talker.py
"""第 02 课：发布者（talker）。

以 1 Hz 频率向话题 /chatter 发布 std_msgs/msg/String 消息。

运行:
    ros2 run talker_listener talker
"""

import rclpy
from rclpy.node import Node
from std_msgs.msg import String


class Talker(Node):

    def __init__(self):
        super().__init__('talker')

        # create_publisher(消息类型, 话题名, QoS 深度)
        # 话题名不写前导 '/' 时为「相对名」，实际解析为 /chatter。
        self._pub = self.create_publisher(String, 'chatter', 10)

        self._count = 0
        self._timer = self.create_timer(1.0, self._on_timer)
        self.get_logger().info('Talker 已上线，正在向 /chatter 发布消息……')

    def _on_timer(self):
        msg = String()
        msg.data = f'Hello from talker #{self._count}'
        self._pub.publish(msg)
        self.get_logger().info(f'发布: {msg.data}')
        self._count += 1


def main(args=None):
    rclpy.init(args=args)
    node = Talker()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()
